File: katas/typescript/src/CleanCode/functions/functions-bad.ts
// Violación de Clean Code: Funciones Mal Diseñadas
// ❌ Problemas: Hacen muchas cosas, grandes, niveles de abstracción mezclados, efectos secundarios

class UserScoreProcessor {
  // ❌ Función hace MUCHAS cosas: calcula bonus, actualiza score, envía notificación y loguea
  // ❌ Niveles de abstracción mezclados
  public processUserScore(user: { name: string; score: number }, newScore: number, timeInSeconds: number): void {
    console.log(`Procesando puntaje para ${user.name}...`);

    // Nivel bajo: cálculo de bonus
    let bonus = 0;
    if (timeInSeconds < 60) {
      bonus = 10;
      console.log(`Aplica bonus de ${bonus} puntos por tiempo rápido.`);
    }

    const finalScore = user.score + newScore + bonus;

    // Nivel medio: actualizar el score del usuario (efecto secundario)
    user.score = finalScore;

    // Nivel bajo: enviar notificación
    console.log(`Notificación enviada a ${user.name}: Tu nuevo puntaje es ${finalScore}`);

    // Nivel bajo: guardar en log
    console.log(`LOG: ${user.name} obtuvo ${newScore} puntos. Puntaje final: ${finalScore}`);
  }
}

// Uso
console.log("=== Violación de Funciones en Clean Code ===");
const user = { name: "Alex", score: 100 };
const processor = new UserScoreProcessor();
processor.processUserScore(user, 50, 30);
console.log("Puntaje final del usuario:", user.score);

export { UserScoreProcessor };

File: katas/python/CleanCode/functions/functions_bad.py
# Violación de Clean Code: Funciones Mal Diseñadas
# ❌ Problemas: Hacen muchas cosas, grandes, niveles de abstracción mezclados, efectos secundarios

from typing import Dict


class UserScoreProcessor:
    # ❌ Función hace MUCHAS cosas: calcula bonus, actualiza score, envía notificación y loguea
    # ❌ Niveles de abstracción mezclados
    def process_user_score(self, user: Dict[str, any], new_score: int, time_in_seconds: int) -> None:
        print(f"Procesando puntaje para {user['name']}...")

        # Nivel bajo: cálculo de bonus
        bonus = 0
        if time_in_seconds < 60:
            bonus = 10
            print(f"Aplica bonus de {bonus} puntos por tiempo rápido.")

        final_score = user["score"] + new_score + bonus

        # Nivel medio: actualizar el score del usuario (efecto secundario)
        user["score"] = final_score

        # Nivel bajo: enviar notificación
        print(
            f"Notificación enviada a {user['name']}: Tu nuevo puntaje es {final_score}")

        # Nivel bajo: guardar en log
        print(
            f"LOG: {user['name']} obtuvo {new_score} puntos. Puntaje final: {final_score}")


if __name__ == "__main__":
    # Uso
    print("=== Violación de Funciones en Clean Code ===")
    user = {"name": "Alex", "score": 100}
    processor = UserScoreProcessor()
    processor.process_user_score(user, 50, 30)
    print(f"Puntaje final del usuario: {user['score']}")

    # ❌ Si necesito cambiar eat() o sleep(), debo modificar 3 clases
    # ❌ Si agrego más animales, más duplicación
    # ❌ Propenso a inconsistencias
